[BE-4443-4405]: Added new test component docs for react native (#684)
appcircleio/appcircle-docusaurus: docs/workflows/react-native-specific-workflow-steps/react-native-unit-test.md

---
title: React Native Unit Test
description: Learn how to run unit tests for your React Native projects easily with Appcircle, ensuring high-quality code and improved app performance.
tags: [react native, mobile, workflow, step, unit, tests, unit test]
---

import Screenshot from '@site/src/components/Screenshot';

# React Native Unit Test

This component runs all the unit tests in your project written with [Jest](https://jestjs.io/docs/tutorial-react-native) integration. When this step is completed, it generates a test report file in `JUnit.xml` format. You can view these test results in detail using Appcircle's Test Report component.

> For detailed information for Continuous Testing. Please visit our [React Native Continuous Testing documentation](/continuous-testing/react-native-testing/react-native-unit-testing).

Reviewer: ```suggestion
For detailed information for continuous testing. Please visit our [React Native Continuous Testing documentation](/continuous-testing/react-native-testing/react-native-unit-testing).
```

To generate detailed Test Reports. Please visit our [Test Reports Component documentation](/workflows/react-native-specific-workflow-steps/test-reports-react-native).

### Prerequisites

The workflow steps that need to be executed before running the **React Native Unit Test** workflow step, along with their respective reasons, are listed in the table below.

| Prerequisite Workflow Step                                                                                | Description                                                                                                                                                                                        |
| --------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | [**Git Clone**](https://docs.appcircle.io/workflows/common-workflow-steps#git-clone)                      | Clone the selected repository to the build machine. Please use the [**Install Node**](https://docs.appcircle.io/workflows/react-native-specific-workflow-steps#install-node) step after this step. |

Reviewer: ```suggestion
| [**Git Clone**](https://docs.appcircle.io/workflows/common-workflow-steps#git-clone)                      | Clone the selected repository to the build machine. |
```

> | [**Install Node**](https://docs.appcircle.io/workflows/react-native-specific-workflow-steps#install-node) | This step will install Node modules for your application. Please note that the **NPM/Yarn Commands** step should be used after this step.                                                          |

Reviewer: ```suggestion
| [**Install Node**](https://docs.appcircle.io/workflows/react-native-specific-workflow-steps#install-node) | This step will install Node modules for your application.                                                           |
```

|--------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [**Git Clone**](/workflows/common-workflow-steps#git-clone)                                | Clone the selected repository to the build machine.                                                                                                                                    |
| [**Install Node**](/workflows/react-native-specific-workflow-steps#install-node)           | This step will install Node modules for your application.                                                                                                                              |
| [**NPM/Yarn Commands**](/workflows/react-native-specific-workflow-steps/npm-yarn-commands) | This step install the [NPM](https://www.npmjs.com/) or [Yarn](https://www.npmjs.com/package/yarn) package manager to install specific dependencies for your React Native applications. |

> <Screenshot url='https://cdn.appcircle.io/docs/assets/BE4443-rnUnitFlowOrder.png'/>

Reviewer 2: We should update the screenshot since the component description got updates while PR was in review.

<img width="1018" alt="Screenshot 2024-11-02 at 08 26 10" src="https://github.com/user-attachments/assets/bcb9f7c9-6ac7-4853-a4cd-949749032c41">

Author: I updated all related screenshots for both UI and Unit test comps @csonuryilmaz

:::danger Step Rule

There's one important setting that you should change. If any workflow steps fail, Appcircle automatically skips other steps and jumps directly to the **Export Build Artifacts** step. However, it is possible that some of your tests may fail. **If React Native Unit Test doesn't run, not all tests will be completed and test results will not be generated.** You should turn on the following toggles so that **React Native Unit Test** component always runs whether your tests fail or pass.

- Always run this step even if the previous steps fail to **ON**
- Continue with the next step even if this step fails to **ON**

> <Screenshot url='https://cdn.appcircle.io/docs/assets/unitRule.png' />

Reviewer 2: Are we sure about this caution box? For example, is it reasonable to "always run unit tests" when, for example, the "npm install" step fails?

As far as I see from your test build profile, you also did not obey your caution.  😅 

<img width="1018" alt="Screenshot 2024-11-02 at 08 34 19" src="https://github.com/user-attachments/assets/406bd65d-453a-4dbb-bedf-0520a4a203cc">

IMO, confusion happened here. Test reports must run when unit/UI tests fail, but the unit/UI test step has different constraints.

Author: Yes you are right, I updated the screenshot and rule according to this. You can check the related screen shots for both comps

:::

### Input Variables

This step contains different variables. It needs these variables to work. The table below gives explanations of these variables.

> <Screenshot url='https://cdn.appcircle.io/docs/assets/BE4443-rnUnitFlowInputs.png' />

Reviewer 2: "Output Direction" seems to be a new vocabulary that has been introduced within react native work.

![Screenshot 2024-11-02 at 08 49 05](https://github.com/user-attachments/assets/df62a38d-232c-40e3-8627-1368c8ae81ce)

As a best practice, we should name similar inputs with similar titles between the components. ⚠️ 

"Output Directory Path" is the most commonly used title for this purpose.

I'm fixing it in the components repository now. Please refresh the screenshot here.

Also, focusing on the "inputs" section in the screenshot rather than the top of the page will be more useful since the purpose of this screenshot is to show the inputs to the user.

<Screenshot url='https://cdn.appcircle.io/docs/assets/BE4443-rnUnitFlowOrder.png'/>

:::danger Step Rule

> There's one important setting that you should change. If any workflow steps fail, Appcircle automatically skips other steps and jumps directly to the Export Build Artifacts step. However, it is possible that some of your tests may fail. **If Test Report Component doesn't run, reports will not be generated.** You should turn on the following toggles so that Test Report Component always runs whether your tests fail or pass.

Reviewer: I have two feedback for this warn:

Can you applied the my suggestion for this warn on Test Reports component document?
IMO this warn must be for React Native UI Test component, not for Test Reports
Already Test Reports has own warn message.

Author: Same rule available for unit test, I update the sentence and the screenshot.

:::danger Step Rule

> There's one important setting that you should change. If any workflow steps fail, Appcircle automatically skips other steps and jumps directly to the Export Build Artifacts step. However, it is possible that some of your tests may fail. **If React Native Unit Test doesn't run, test results will not be generated.** You should turn on the following toggles so that React Native Unit Test Component always runs whether your tests fail or pass.

Reviewer: ```suggestion
There's one important setting that you should change. If any workflow steps fail, Appcircle automatically skips other steps and jumps directly to the **Export Build Artifacts** step. However, it is possible that some of your tests may fail. **If React Native Unit Test doesn't run, not all tests will be completed and test results will not be generated.** You should turn on the following toggles so that **React Native Unit Test** component always runs whether your tests fail or pass.
```

The outputs resulting from the operation of this component are as follows:

| Variable Name         | Description                                                                                                     |
|-----------------------|-----------------------------------------------------------------------------------------------------------------|
| `AC_TEST_RESULT_PATH` | The output path for the `JUnit-report.xml` file. This environment variable can be utilized in subsequent steps. |

:::info Coverage Support

> Appcircle does **not** currently support **coverage** calculation in React Native projects. Coverage calculation for test reports in React Native projects is available in our development **roadmap** for this component. It will be available **as soon as** it is available.

Reviewer: It is not "must to" only "can be":
```suggestion
Appcircle does **not** currently support **test coverage** calculation for React Native projects. However, adding this feature is part of our development **roadmap** and will be included in a future update. We will make it available **as soon as** development is completed.
```
